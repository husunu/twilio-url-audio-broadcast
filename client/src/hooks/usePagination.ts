import { useState, useCallback, useMemo, useEffect } from 'react';
import { useRowsPerPageOptions } from '@/hooks/useRowsPerPageOptions';

/**
 * All around hook for paginating a collection of data client side.
 * @param data List of elements to paginate
 */
export const usePagination = (data: any[]) => {
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(10);

  const rowsPerPageOptions = useRowsPerPageOptions(data.length);

  const handleChangePage = useCallback(
    (event, newPage) => {
      setPage(newPage);
    },
    [setPage]
  );

  const handleChangeRowsPerPage = useCallback(
    (event) => {
      setRowsPerPage(parseInt(event.target.value, 10));
      setPage(0);
    },
    [setPage, setRowsPerPage]
  );

  const paginatedData = useMemo(() => {
    return data.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage);
  }, [data, page, rowsPerPage]);

  useEffect(() => {
    setPage(0);
  }, [data]);

  return {
    page,
    rowsPerPage,
    handleChangePage,
    handleChangeRowsPerPage,
    paginatedData,
    rowsPerPageOptions,
  };
};
