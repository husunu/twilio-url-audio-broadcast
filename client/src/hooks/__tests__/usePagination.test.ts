import { renderHook, act, cleanup } from '@testing-library/react-hooks';
import { usePagination } from '../usePagination';

const numbers = [...Array(30).keys()];

let data: number[];

describe('usePagination', () => {
  beforeEach(() => {
    data = numbers;
  });

  afterEach(() => {
    cleanup();
  });

  test('empty array', () => {
    let { result } = renderHook(() => usePagination(data));
    expect(result).toMatchSnapshot();
  });

  test('handleChangePage', () => {
    let { result } = renderHook(() => usePagination(data));

    act(() => {
      result.current.handleChangePage(null, 2);
    });

    expect(result.current).toMatchSnapshot();
  });

  test('handleChangeRowsPerPage', () => {
    let { result } = renderHook(() => usePagination(data));

    act(() => {
      result.current.handleChangeRowsPerPage({ target: { value: 25 } });
    });

    expect(result.current).toMatchSnapshot();
  });

  test('page goes back to 0 every time data changes', () => {
    let { result, rerender } = renderHook((data) => usePagination(data), {
      initialProps: [...numbers],
    });

    act(() => {
      result.current.handleChangePage(null, 1);
      rerender([]);

      expect(result.current.page).toBe(0);
    });
  });
});
