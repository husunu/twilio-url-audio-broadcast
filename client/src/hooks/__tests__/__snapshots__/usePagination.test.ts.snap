// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePagination empty array 1`] = `
Object {
  "current": Object {
    "handleChangePage": [Function],
    "handleChangeRowsPerPage": [Function],
    "page": 0,
    "paginatedData": Array [
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
    ],
    "rowsPerPage": 10,
    "rowsPerPageOptions": Array [
      10,
      25,
    ],
  },
  "error": undefined,
}
`;

exports[`usePagination handleChangePage 1`] = `
Object {
  "handleChangePage": [Function],
  "handleChangeRowsPerPage": [Function],
  "page": 2,
  "paginatedData": Array [
    20,
    21,
    22,
    23,
    24,
    25,
    26,
    27,
    28,
    29,
  ],
  "rowsPerPage": 10,
  "rowsPerPageOptions": Array [
    10,
    25,
  ],
}
`;

exports[`usePagination handleChangeRowsPerPage 1`] = `
Object {
  "handleChangePage": [Function],
  "handleChangeRowsPerPage": [Function],
  "page": 0,
  "paginatedData": Array [
    0,
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8,
    9,
    10,
    11,
    12,
    13,
    14,
    15,
    16,
    17,
    18,
    19,
    20,
    21,
    22,
    23,
    24,
  ],
  "rowsPerPage": 25,
  "rowsPerPageOptions": Array [
    10,
    25,
  ],
}
`;
